plugins/modem-manager/fu-firehose-updater.c

```diff
@@ -0,0 +1,699 @@
+		program_sector_size_in_bytes = xb_node_get_attr_as_uint (node, "SECTOR_SIZE_IN_BYTES");
+		g_warn_if_fail (program_file != NULL);
+		g_warn_if_fail (program_filename != NULL);
+		g_warn_if_fail (program_sector_size_in_bytes != G_MAXUINT64);
+
+		total_bytes += g_bytes_get_size (program_file);
+
+		if (program_sector_size_in_bytes > max_payload_size) {
+			g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+				     "Failed to validate program file '%s' command: "
+				     "requested sector size bigger (%" G_GUINT64_FORMAT " bytes) "
+				     "than maximum payload size agreed with device (%u bytes)",
+				     program_filename, program_sector_size_in_bytes, max_payload_size);
+			return FALSE;
+		}
+	}
+
+	for (guint i = 0; i < action_nodes->len; i++) {
+		XbNode *node = g_ptr_array_index (action_nodes, i);
+		const gchar *action;
+		gchar *cmd_str = NULL;
+		gboolean rawmode = FALSE;
+		GBytes *program_file = NULL;
+		const gchar *program_filename = NULL;
+		guint64 program_sector_size = 0;
+		guint payload_size = 0;
+		GByteArray *cmd_bytearray = NULL;
+		g_autoptr(XbSilo) rsp_silo = NULL;
+		g_autoptr(XbNode) rsp_node = NULL;
+
+		action = xb_node_get_element (node);
+
+		cmd_str = xb_node_export (node, XB_NODE_EXPORT_FLAG_COLLAPSE_EMPTY, error);
+		if (cmd_str == NULL)
+			return FALSE;
+		cmd_bytearray = g_byte_array_new_take ((guint8 *)cmd_str, strlen (cmd_str));
+
+		g_debug ("running command '%s'...", action);
+		if (!fu_firehose_updater_send_and_receive (self, cmd_bytearray, &rsp_silo, &rsp_node, error)) {
+			g_prefix_error (error, "Failed to run command '%s': ", action);
+			return FALSE;
+		}
+
+		if (!fu_firehose_updater_check_operation_result (rsp_node, &rawmode)) {
+			g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+				     "Command '%s' failed", action);
+			return FALSE;
+		}
+
+		if (g_strcmp0 (action, "program") != 0)
+			continue;
+
+		program_file = xb_node_get_data (node, "fwupd:ProgramFile");
+		program_filename = xb_node_get_attr (node, "filename");
+		program_sector_size = xb_node_get_attr_as_uint (node, "SECTOR_SIZE_IN_BYTES");
+		g_warn_if_fail (program_file != NULL);
+		g_warn_if_fail (program_filename != NULL);
+		g_warn_if_fail (program_sector_size != G_MAXUINT64);
+
+		if (rawmode == FALSE) {
+			g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+				     "Failed to download program file '%s': rawmode not enabled",
+				     program_filename);
+			return FALSE;
+		}
+
+		while ((payload_size + (guint)program_sector_size) < max_payload_size)
+			payload_size += (guint)program_sector_size;
+
+		g_debug ("sending program file '%s' (%zu bytes)", program_filename, g_bytes_get_size (program_file));
+		if (!fu_firehose_updater_send_program_file (self, program_filename, program_file,
+							    payload_size, program_sector_size, error)) {
+			g_prefix_error (error, "Failed to send program file '%s': ",
+					program_filename);
+			return FALSE;
+		}
+
+		g_clear_object (&rsp_node);
+		g_clear_object (&rsp_silo);
```

Comment:
this is normally an indication we should split this up as well...

Reply:
refactored into several function, should look good now